File: frontend/packages/types/src/data/actor.ts
export enum ActorKind {
  user = 'user',
  footprint = 'footprint',
  firmEmployee = 'firm_employee',
  organization = 'organization',
  apiKey = 'api_key',
}

export type ActorUser = {
  kind: ActorKind.user;
  fpId: string;
};

export type ActorFootprint = {
  kind: ActorKind.footprint;
};

export type ActorFirmEmployee = {
  kind: ActorKind.firmEmployee;
};

export type ActorOrganization = {
  kind: ActorKind.organization;
  member: string;
  firstName?: string;
  lastName?: string;
  email: string;
};

export type ActorApiKey = {
  kind: ActorKind.apiKey;
  id: string;
  name: string;
};

export type Actor = ActorUser | ActorFootprint | ActorFirmEmployee | ActorOrganization | ActorApiKey;

File: backend/components/api/wire_types/src/resources/actor.rs
use crate::*;
use newtypes::FpId;
use newtypes::TenantApiKeyId;

#[derive(Debug, Clone, Serialize, Apiv2Schema)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "kind")]
pub enum Actor {
    /// A user onboarding with Footprint.
    User { id: FpId },
    /// A Footprint automated process
    Footprint,
    /// A Footprint employee via the dashboard
    FirmEmployee,
    /// A dashboard user for a tenant
    Organization {
        member: String,
        email: String,
        first_name: Option<String>,
        last_name: Option<String>,
    },
    /// A tenant API key
    ApiKey { id: TenantApiKeyId, name: String },
}
